Add initial support for ICUs in template pipeline

mmalerba opened the pull request:
See individual commit messages for details.

mmalerba committed:
refactor(compiler): Add support for ingesting ICUs

ICUs are now ingested by adding ops to both the creation and update IR.
Both of these ops are ultimately removed before reification, but they
are needed to coordinate and link data between the creation and update
ops. This is done in a new ICU extraction phase that removes both ICU
ops and adds an i18nExpr op to the update IR.

mmalerba committed:
refactor(compiler): Resolve ICU params during i18n post-processing

ICU params in i18n messages are now resolved in the post-processing call
rather than in the initial message creation. This matches the output
generated by TemplateDefinitionBuilder.

mmalerba committed:
refactor(compiler): Wrap bare ICUs in an i18n block

ICUs can be used outside of an i18n block. In this case the ICU should
be automatically wrapped in a new i18n block. This commit adds a new
phase to handle wrapping these bare ICUs.

dylhunn commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -383,6 +385,16 @@ function ingestDeferBlock(unit: ViewCompilationUnit, deferBlock: t.DeferredBlock
+function ingestIcu(unit: ViewCompilationUnit, icu: t.Icu) {
+  if (icu.i18n instanceof i18n.Message) {
+    const xref = unit.job.allocateXrefId();
+    unit.create.push(ir.createIcuOp(xref, icu.i18n, null!));
```
Just a heads up, i18n spans have tests in the source spans unit test folder. I don't blame you for not wanting to deal with that at this point though.

mmalerba replied:
Yeah, I was thinking maybe I could do i18n (except source maps) and you could do source maps (except i18n) and then we can team up and use our combined knowledge to take down the combo

dylhunn replied:
Sounds scary! Non-i18n source maps are more or less done, so we'll team up when you're ready and find out how they interact... 👻

dylhunn commented on packages/compiler/src/template/pipeline/ir/src/enums.ts:
```diff
@@ -426,3 +426,18 @@ export enum BindingKind {
+ */
+export enum I18nParamResolutionTime {
+  /**
+   * Param is resolved at message creation time.
```
Can you add some details about why/which placeholders get resolved at creation vs update time? Or is it just truly arbitrary?

mmalerba replied:
👍

dylhunn commented on packages/compiler/src/template/pipeline/src/phases/wrap_icus.ts:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Wraps ICUs that do not already belong to an i18n block in a new i18n block.
+ */
+export function phaseWrapIcus(job: CompilationJob): void {
```
Makes sense. Weird `i18n` is not required though.

mmalerba replied:
I guess its slightly more ergonomic for users to just be able to throw an ICU in without needing to add an attribute 🤷

mmalerba committed:
docs(compiler): Clarify the purpose of i18n param resolution time

Adds some additional doc comments clarifying when the `Creation`
resolution time is used vs the `Postprocessing` resolution time.
